File: frontend/src/types/notification.ts
// 인앱 알림 타입 — 백엔드 domain/notification 과 1:1 대응.

import type { PageMeta } from '@/types/page';

/**
 * 알림 종류. 서버 NotificationType enum 과 같은 값이다.
 *
 * <b>지금 서버가 만드는 알림은 이 셋뿐이다.</b> 알림 수신 설정(GET /members/me/notification-settings)
 * 에는 커뮤니티·문의·공지·랭킹 항목도 있지만, 그 알림을 저장하는 코드는 아직 없다.
 * 그래서 목록에는 나타나지 않는다.
 *
 * 이번 범위에서 화면이 이동까지 연결하는 것은 `FRIENDSHIP`·`DM` 둘이다.
 * `REPORT` 는 목록에 보여주고 읽음 처리만 한다.
 */
export type NotificationType = 'FRIENDSHIP' | 'DM' | 'REPORT';

/**
 * 알림이 가리키는 대상의 종류. `referenceId` 를 무엇으로 읽어야 하는지 알려 준다.
 *
 * - `FRIENDSHIP` → `referenceId` 는 `friendshipId`
 * - `DM_ROOM`    → `referenceId` 는 DM `roomId`
 *
 * 그 외 값은 계약에 정의돼 있지 않아 좁히지 않는다. 아는 값만 분기하고 나머지는 이동 없이 둔다.
 */
export type ReferenceType = string;

export interface NotificationItem {
  notificationId: number;
  type: NotificationType;
  title: string;
  message: string;
  referenceType: ReferenceType;
  referenceId: number | null;
  /** 읽은 시각. null 이면 아직 안 읽은 알림이다 */
  readAt: string | null;
  createdAt: string;
}

/** GET /notifications — 최신순 */
export interface NotificationListResponse {
  notifications: NotificationItem[];
  page: PageMeta;
}

/**
 * `FRIENDSHIP` 알림은 <b>"친구 신청을 받았다"만</b> 뜻한다.
 *
 * 신청이 수락됐을 때는 서버가 알림을 만들지 않는다(2026-08-04 확정). 두 경우가 같은
 * `type`·`referenceType` 을 쓰던 탓에 제목 문자열로 구분해야 했는데, 수락 알림 자체를
 * 없애서 그 문제를 지웠다. 덕분에 이 타입의 알림은 언제나 '받은 요청' 화면으로 보내면 된다.
 */
export const FRIENDSHIP_REFERENCE_TYPE = 'FRIENDSHIP';

/** DM 알림의 `referenceType`. `referenceId` 가 방 번호다. */
export const DM_ROOM_REFERENCE_TYPE = 'DM_ROOM';

File: backend/src/main/java/com/protractor/backend/domain/notification/entity/Notification.java
package com.protractor.backend.domain.notification.entity;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import java.time.LocalDateTime;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;

/** 회원에게 도착한 알림 한 건. 공식 스키마 {@code notifications} 매핑. */
@Entity
@Table(name = "notifications")
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Notification {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "notification_id")
    private Long id;

    @Column(name = "receiver_member_id", nullable = false)
    private Long receiverMemberId;

    /** 시스템 발송(리포트 안내 등)은 보낸 사람이 없다. 친구 신청처럼 사람이 보내는 알림에서만 채운다. */
    @Column(name = "sender_member_id")
    private Long senderMemberId;

    @Enumerated(EnumType.STRING)
    @Column(name = "type", nullable = false, length = 50)
    private NotificationType type;

    @Column(name = "title", nullable = false, length = 200)
    private String title;

    @Column(name = "message", length = 500)
    private String message;

    /** 클릭 시 이동할 대상의 종류·id. 리포트 안내 알림은 발송 시점에 리포트가 없을 수 있어 id가 비어 있다. */
    @Column(name = "reference_type", length = 50)
    private String referenceType;

    @Column(name = "reference_id")
    private Long referenceId;

    /** 비어 있으면 아직 안 읽은 알림이다. */
    @Column(name = "read_at")
    private LocalDateTime readAt;

    @CreationTimestamp
    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    private Notification(Long receiverMemberId, NotificationType type, String title, String message,
            String referenceType, Long referenceId) {
        this.receiverMemberId = receiverMemberId;
        this.type = type;
        this.title = title;
        this.message = message;
        this.referenceType = referenceType;
        this.referenceId = referenceId;
    }

    /** 주간 리포트 안내 알림. */
    public static Notification weeklyReport(Long receiverMemberId, String title, String message) {
        return new Notification(receiverMemberId, NotificationType.REPORT, title, message,
                NotificationType.REPORT.name(), null);
    }

    public static Notification friendship(Long receiverMemberId, Long senderMemberId, Long friendshipId,
            String title, String message) {
        Notification notification = new Notification(receiverMemberId, NotificationType.FRIENDSHIP, title, message,
                "FRIENDSHIP", friendshipId);
        notification.senderMemberId = senderMemberId;
        return notification;
    }

    public static Notification dm(Long receiverMemberId, Long senderMemberId, Long roomId, String message) {
        Notification notification = new Notification(receiverMemberId, NotificationType.DM, "새 DM 메시지", message,
                "DM_ROOM", roomId);
        notification.senderMemberId = senderMemberId;
        return notification;
    }

    /** 읽음 처리. 이미 읽은 알림은 처음 읽은 시각을 유지한다(같은 요청을 다시 보내도 결과가 같다). */
    public void markRead() {
        if (this.readAt == null) {
            this.readAt = LocalDateTime.now();
        }
    }
}
